File: packages/deep-engine-native/src/runtime_package/material_bindings.rs
use std::collections::{HashMap, HashSet};

use crate::{contract::RenderPacket, shader_package::DeepShaderPackageV2};

use super::{RuntimeMaterialShaderBinding, RuntimePackageError, fail};

pub(super) fn validate(
    bindings: &[RuntimeMaterialShaderBinding],
    packet: &RenderPacket,
    shaders: &[DeepShaderPackageV2],
) -> Result<(), RuntimePackageError> {
    if bindings.is_empty() || bindings.len() > 16_384 {
        return fail("runtime package v2 requires 1..=16384 material bindings");
    }
    let materials: HashSet<_> = packet
        .materials
        .iter()
        .map(|value| value.id.as_str())
        .collect();
    let shaders: HashMap<_, _> = shaders
        .iter()
        .map(|value| (value.package_id.as_str(), value))
        .collect();
    let mut used = HashSet::new();
    let mut previous: Option<&str> = None;
    for binding in bindings {
        let material = binding.material_id.as_str();
        if material.is_empty()
            || material.len() > 256
            || binding.technique_id.is_empty()
            || binding.technique_id.len() > 128
        {
            return fail("invalid material binding identity or technique");
        }
        if previous.is_some_and(|id| id >= material) {
            return fail("material bindings must be sorted by unique material id");
        }
        previous = Some(material);
        if !materials.contains(material) {
            return fail(format!(
                "material binding refers to absent material {material:?}"
            ));
        }
        let Some(shader) = shaders.get(binding.package_id.as_str()) else {
            return fail(format!(
                "material {material:?} refers to an absent shader package"
            ));
        };
        if shader.shader_abi.id != "deep.pbr.mesh.v2" {
            return fail("executable materials require CSM shader ABI deep.pbr.mesh.v2");
        }
        if !shader
            .passes
            .iter()
            .any(|pass| pass.technique_id == binding.technique_id && pass.kind == "forward")
        {
            return fail(format!(
                "material {material:?} technique has no forward pass"
            ));
        }
        used.insert(binding.package_id.as_str());
    }
    if used.len() != shaders.len() {
        return fail("every shader entrypoint must be bound to a material");
    }
    Ok(())
}

File: packages/deep-engine/src/runtimePackage/materialBindings.ts
import type { DeepShaderPackageV2 } from "../shaderPackage/types.js";
import { compareRuntimeStrings } from "./hash.js";
import { array, fields, record, requireValue, resourceId, snapshotJson, string } from "./primitives.js";
import type { RuntimeMaterialShaderBinding } from "./types.js";

function readBindings(input: unknown, allowEmpty = false): RuntimeMaterialShaderBinding[] {
  const entries = array(input, "$.materialBindings", 16_384);
  requireValue(allowEmpty || entries.length > 0, "$.materialBindings", "Runtime package v2 requires material bindings.");
  return entries.map((candidate, index) => {
    const path = `$.materialBindings[${index}]`, entry = record(candidate, path);
    fields(entry, ["materialId", "packageId", "techniqueId"], [], path);
    const materialId = string(entry.materialId, `${path}.materialId`);
    const techniqueId = string(entry.techniqueId, `${path}.techniqueId`);
    requireValue(materialId.length > 0 && new TextEncoder().encode(materialId).length <= 256, path, "Invalid material identity.");
    requireValue(techniqueId.length > 0 && new TextEncoder().encode(techniqueId).length <= 128, path, "Invalid technique identity.");
    return { materialId, packageId: resourceId(entry.packageId, `${path}.packageId`), techniqueId };
  });
}

/** Owns author input and orders Unicode material ids identically to the native UTF-8 comparator. */
export function normalizeRuntimeMaterialBindings(input: unknown, allowEmpty = false): readonly RuntimeMaterialShaderBinding[] {
  return readBindings(snapshotJson(input), allowEmpty).sort((a, b) => compareRuntimeStrings(a.materialId, b.materialId));
}

export function validateRuntimeMaterialBindings(input: unknown, materials: readonly { readonly id: string }[],
  shaders: ReadonlyMap<string, DeepShaderPackageV2>, allowEmpty = false): void {
  const bindings = readBindings(input, allowEmpty), materialIds = new Set(materials.map(material => material.id)), used = new Set<string>();
  let previous: string | undefined;
  for (const [index, binding] of bindings.entries()) {
    const path = `$.materialBindings[${index}]`;
    requireValue(previous === undefined || compareRuntimeStrings(previous, binding.materialId) < 0,
      path, "Material bindings must be sorted by unique material id.");
    previous = binding.materialId;
    requireValue(materialIds.has(binding.materialId), path, "Material binding refers to an absent material.");
    const shader = shaders.get(binding.packageId);
    requireValue(shader, path, "Material binding refers to an absent shader package.");
    requireValue(String(shader.shaderAbi.id) === "deep.pbr.mesh.v2", path, "Executable materials require the CSM shader ABI deep.pbr.mesh.v2.");
    requireValue(shader.passes.some(pass => pass.techniqueId === binding.techniqueId && pass.kind === "forward"),
      path, "Material binding technique has no forward pass.");
    used.add(binding.packageId);
  }
  requireValue(used.size === shaders.size, "$.materialBindings", "Every shader entrypoint must be bound to a material.");
}
